### ISSUE TITLE
CampaignMember Duplicate Error when adding Closed Won and Closed Lost Opp

### ISSUE THREAD

**Issue by IanShoe** [SalesforceFoundation/Cumulus#2599]
If I try to insert 2 Opps in the same list with the same npsp__Primary_Contact__c and campaign but one with a Closed Won status and the other Closed Lost, NPSP tries to add the contact twice to a campaign.

The use case I have is an integration with Eventbrite and someone issues a refund but recreates a new order that ultimately ends up as 2 opportunities. We process these in nightly batches.

Below is code that replicates the error.

```
Campaign camp = new Campaign();
camp.isactive = true;
camp.name = 'Campaign';
insert camp;

Account acc = new Account();
acc.name = 'Account';
insert acc;

Contact con = new Contact();
con.lastName = 'Contact';
con.accountid = acc.id;
insert con;

Opportunity opp1 = new Opportunity();
opp1.amount = 10;
opp1.stagename = 'Closed Won';
opp1.closeDate = Date.today();
opp1.AccountId = acc.id;
opp1.CampaignId = camp.id;
opp1.npsp__Primary_Contact__c = con.id;
opp1.name = 'one';

Opportunity opp2 = new Opportunity();
opp2.amount = 10;
opp2.stagename = 'Closed Lost';
opp2.closeDate = Date.today();
opp2.AccountId = acc.id;
opp2.CampaignId = camp.id;
opp2.npsp__Primary_Contact__c = con.id;
opp2.name = 'two';

List<Opportunity> opps = new List<Opportunity>();
opps.add(opp1);
opps.add(opp2);

insert opps;
```

Below is the error:

Line: 37, Column: 1
System.DmlException: Insert failed. First exception on row 0; first error: CANNOT_INSERT_UPDATE_ACTIVATE_ENTITY, npsp.TDTM_Opportunity: execution of AfterInsert caused by: System.DmlException: Insert failed. First exception on row 1; first error: DUPLICATE_VALUE, Attempted to add an entity '0036300000IxccF' to a campaign '701630000002ZvA' more than once.: [] (npsp) : []

**Comment by davidhabib** [comment]
**lurch: add

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x80000006oYitAAE" target="_blank">W-018990</a>

**Comment by IanShoe** [comment]
@davidhabib Just curious if this is confirmed as a bug and if there's a fix in the works.

**Comment by judisohn** [comment]
Yes, it's confirmed but we haven't yet placed this in a sprint for developer attention. Thanks for your patience.

**Comment by judisohn** [comment]
Another [thread](https://powerofus.force.com/0D58000003VqTri).

**Comment by judisohn** [comment]
Another [thread](https://powerofus.force.com/0D51E00004njaHQ).

**Comment by jamesbsfdc** [comment]
Another [thread](https://powerofus.force.com/s/feed/0D51E00005PBct6).

**Comment by mbeaty-sf** [comment]
W-7782881

**Comment by kporray1** [comment]
Please follow this [Known Issue](https://trailblazer.salesforce.com/issues_view?id=a1p4V000002dlFyQAI&title=campaignmember-duplicate-error-when-adding-closed-won-and-closed-lost-opportunities) in the Trailblazer Community for updates.
